### PR TITLE
xcatprobe osdeploy performance

### PR DESCRIPTION
**Pull request by xuweibj**
Support value: 'origin'  'phase' 'compact'
Support diskful

```
# xcatprobe -w osdeploy -n c910f03c17k20 -p compact
The install NIC in current server is enp0s1                                                                       [INFO]
All nodes to be deployed are valid                                                                                [ OK ]
-------------------------------------------------------------
Start capturing every message during OS provision process....
-------------------------------------------------------------
.........
==================osdeploy_probe_report=================
All nodes provisioned successfully                                                                                [ OK ]
----------------------------
|node provision performance|
----------------------------
| NODE          | ELAPSED  |
| c910f03c17k20 | 00:08:54 |
| c910f03c17k20 | 00:08:59 |
----------------------------
```

```
# xcatprobe -w osdeploy -n c910f03c17k20 -p phase
The install NIC in current server is enp0s1                                                                       [INFO]
All nodes to be deployed are valid                                                                                [ OK ]
-------------------------------------------------------------
Start capturing every message during OS provision process....
-------------------------------------------------------------
.....
==================osdeploy_probe_report=================
All nodes provisioned successfully                                                                                [ OK ]
------------------------------------------------------------------------
|                      node provision performance                      |
------------------------------------------------------------------------
| NODE          | SVRBOOT  | RPM      | POST     | POSTBOOT | ELAPSED  |
| c910f03c17k20 | 00:00:08 | 00:02:27 | 00:00:19 | 00:00:00 | 00:08:54 |
| c910f03c17k21 | 00:00:08 | 00:02:28 | 00:00:20 | 00:00:00 | 00:08:59 |
------------------------------------------------------------------------
```

```
# xcatprobe -w osdeploy -n c910f03c17k20 -p origin
The install NIC in current server is enp0s1                                                                       [INFO]
All nodes to be deployed are valid                                                                                [ OK ]
-------------------------------------------------------------
Start capturing every message during OS provision process....
-------------------------------------------------------------
.....
==================osdeploy_probe_report=================
All nodes provisioned successfully                                                                                [ OK ]
-----------------------------------------------------------------------------------------------------------------------
|                                             node provision performance                                              |
-----------------------------------------------------------------------------------------------------------------------
| NODE          | RPOWER   | DHCP     | BOOTLOADER | KERNEL   | INITRD   | RPM      | POST     | POSTBOOT | COMPLETED |
| c910f03c17k20 | 21:58:27 | 21:58:35 | 21:58:52   | 21:58:52 | 21:59:00 | 22:00:07 | 22:05:29 | 22:07:21 | 22:07:21  |
| c910f03c17k21 | 21:58:27 | 21:58:35 | 21:58:53   | 21:58:52 | 21:59:00 | 22:00:07 | 22:05:29 | 22:07:26 | 22:07:26  |
-----------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

File: xCAT-probe/lib/perl/probe_utils.pm [hunk 1]
```diff
@@ -550,4 +550,52 @@ sub is_ntp_ready{
...
+
+    if ($second_in == 0) {
+        return "00:00:00";
+    }
+
+    my $count = 0;
+    while ($count < 3) {
+        my $tmp_second;
+        if ($count == 2) {
+            $tmp_second = $second_in % 100;
```
Comment (hu-weihua) [inline xCAT-probe/lib/perl/probe_utils.pm]: I can understand why you use 100 here, but I don't think it is a good option. If we really have a extreme big number,  such like ``80:15:30``, customer will very confusing. It is hard to understand. we have 2 options, one is ``80h15m30s``, or just keep the biggest number is ``60:59:59``.

File: xCAT-probe/lib/perl/probe_utils.pm [hunk 2]
```diff
@@ -550,4 +550,52 @@ sub is_ntp_ready{
...
+        } else {
+            $tmp_second = $second_in % 60;
+        }
+        if ($tmp_second == 0) {
+            push @time, "00";
+        } elsif ($tmp_second < 10) {
+            push @time,  "0" . "$tmp_second";
```
Comment (hu-weihua) [inline xCAT-probe/lib/perl/probe_utils.pm]: why we can't combine line 584 and 586? just 

    if($tmp_second < 10) {
         push @time, "0$tmp_second";
    }else{
         push @time, "$tmp_second";
    }

File: xCAT-probe/subcmds/osdeploy [hunk 3]
```diff
@@ -65,6 +65,9 @@ Options:
     -r : Trigger 'Replay history' mode. Follow the duration of rolling back. Units are 'h' (hour) or 'm' (minute)
          Supported format examples: 3h30m (3 hours and 30 minutes ago), 2h (2 hours ago), 40m (40 minutes ago) and 3 (3 hours ago).
          If unit is not specified, hour will be used by default.
+    -p : Performance of provision for each node and all. 
+         Supported level: 1 (show how much time spent for provision), 
+         2 (show how much time spent for DHCP, Download RPM packages, Run Postscripts, Run Postbootscripts).
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: -p Show elapsed time of each stage during provision for each node 
     Support 2 output format:
     1 Elapsed time of each stage during provision
     2 Elapsed time from power on node to each stage starting

Maybe @gurevichmark could help to refine the statement.

File: xCAT-probe/subcmds/osdeploy [hunk 4]
```diff
@@ -484,16 +500,11 @@ sub conclusion_report {
         # 2 When replay, if there isn't reboot operation for target node during the rollback time window 
         #   That means there isn't provision process happened 
 
-        if ($monitor) {
-            if (!$isntalling && $postbootscript) {
-                $failed_node{$node}{non_provision_prediction} = "Target node just reboot from disk";
-                next;
-            }
-        } else {
+        unless ($monitor) {
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: If modify this section. could you update the above comments at the same time.

File: xCAT-probe/subcmds/osdeploy [hunk 5]
```diff
@@ -538,13 +551,146 @@ sub conclusion_report {
...
+        next if ($node eq "all");
+
+        # if did not get start time of power on, use osdeploy start time and show warning info
+        if ((!$node_status_time{$node}{$::STATE_POWER_ON}{start_time}) and $node_status_time{all}) {
+            $node_status_time{$node}{$::STATE_POWER_ON}{start_time} = $node_status_time{all};
+            $warninfo = "Did not get start time from log during running, use xcatprobe start time";
+        }
+
+        if ($all_spend_time{start_time}) {
+            $all_spend_time{start_time} = ($all_spend_time{start_time} <= $node_status_time{$node}{$::STATE_POWER_ON}{start_time}) ?  $all_spend_time{start_time} : $node_status_time{$node}{$::STATE_POWER_ON}{start_time};
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: Can be refine to 

     $all_spend_time{start_time} = $node_status_time{$node}{$::STATE_POWER_ON}{start_time} if ($node_status_time{$node}{$::STATE_POWER_ON}{start_time}<$all_spend_time{start_time} );

File: xCAT-probe/subcmds/osdeploy [hunk 6]
```diff
@@ -538,13 +551,146 @@ sub conclusion_report {
...
+    }
+
+    # calculate time for each node and status
+    foreach my $node (keys %node_status_time) {
+        next if ($node eq "all");
+
+        # if did not get start time of power on, use osdeploy start time and show warning info
+        if ((!$node_status_time{$node}{$::STATE_POWER_ON}{start_time}) and $node_status_time{all}) {
+            $node_status_time{$node}{$::STATE_POWER_ON}{start_time} = $node_status_time{all};
+            $warninfo = "Did not get start time from log during running, use xcatprobe start time";
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: If just didn't get start time for one node, this message will show after the whole table.  In fact, other nodes still use their start time. If we add node name in the ``$warning`` message,  but there are more than one node didn't get start time, the message of the one handled later will replace the message of the one handled before.  In fact, if we need to show such message after the whole table, we need to give which nodes didn't get start time.

File: xCAT-probe/subcmds/osdeploy [hunk 7]
```diff
@@ -538,13 +551,146 @@ sub conclusion_report {
...
+
+        if ($all_spend_time{start_time}) {
+            $all_spend_time{start_time} = ($all_spend_time{start_time} <= $node_status_time{$node}{$::STATE_POWER_ON}{start_time}) ?  $all_spend_time{start_time} : $node_status_time{$node}{$::STATE_POWER_ON}{start_time};
+        } else {
+            $all_spend_time{start_time} = $node_status_time{$node}{$::STATE_POWER_ON}{start_time};
+        }
+
+        if ($node_status_time{$node}{$::STATE_COMPLETED}{start_time}) {
+            $all_spend_time{end_time} = ($all_spend_time{end_time} >= $node_status_time{$node}{$::STATE_COMPLETED}{start_time}) ? $all_spend_time{end_time} : $node_status_time{$node}{$::STATE_COMPLETED}{start_time};
+        }
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: Does not `` $all_spend_time{end_time} `` need to assigned the first value when it didn't have initial value before,  just like `` $all_spend_time{start_time} ``do in  line 618.

File: xCAT-probe/subcmds/osdeploy [hunk 8]
```diff
@@ -821,6 +971,7 @@ sub handle_dhcp_msg {
 
             $ipnodemap{$ip} = $node;
             set_node_state($node_state_ref, $node, $::STATE_DHCP);
+            $node_status_time{$node}{$::STATE_DHCP}{end_time} = $log_ref->{time} if ($performance);
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: 1 You handled DHCP package in PPC environment , but miss half DHCP part in non-PPC environment.   

2  if line 944 has ``!$node_status_time{$node}{$::STATE_DHCP}{end_time}``, maybe line 954 and line 974 should have too.

File: xCAT-probe/subcmds/osdeploy [hunk 9]
```diff
@@ -892,16 +1046,29 @@ sub handle_http_msg {
...
                 push (@{ $node_state_ref->{ $ipnodemap{$ip} }{errors}{$::STATE_KICKSTART} }, "$record failed with $http_code") if ($http_code >= 400);
             } elsif ($file =~ /\.deb$/i or $file =~ /\/Packages\/.+\.rpm$/ or $file =~ /\/suse\/noarch\/.+\.rpm$/i) {
                 set_node_state($node_state_ref, $ipnodemap{$ip}, $::STATE_INSTALLRPM);
+                if ($performance) {
+                    if ($node_status_time{$ipnodemap{$ip}}{$::STATE_INSTALLRPM}{start_time}) {
+                        $node_status_time{$ipnodemap{$ip}}{$::STATE_INSTALLRPM}{end_time} = $log_ref->{time};
+                    } else {
+                        $node_status_time{$ipnodemap{$ip}}{$::STATE_INSTALLRPM}{start_time} = $log_ref->{time};
+                    }
+                }
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: In extreme situation， if we only have one RPM, we will miss ``end_time``.

I think we can refine to
 
    if($performance) {
        $node_status_time{$ipnodemap{$ip}}{$::STATE_INSTALLRPM}{end_time} = $log_ref->{time};
        $node_status_time{$ipnodemap{$ip}}{$::STATE_INSTALLRPM}{start_time} =  $log_ref->{time}  unless ( $node_status_time{$ipnodemap{$ip}}{$::STATE_INSTALLRPM}{start_time} );
    }

File: xCAT-probe/subcmds/osdeploy [hunk 10]
```diff
@@ -1000,8 +1179,14 @@ sub handle_compute_msg {
                     push @{ $node_state_ref->{$node}{errors}{$::STATE_POSTSCRIPT} }, $error_str;
                 }
             }
+            $node_status_time{$node}{$::STATE_POSTSCRIPT}{end_time} = $log_ref->{time} if ($performance);
         } elsif ($log_ref->{msg} =~ /Running postbootscript:/i) {
             set_node_state($node_state_ref, $node, $::STATE_POSTBOOTSCRIPT);
+            if ($performance) {
+                if (!$node_status_time{$node}{$::STATE_POSTBOOTSCRIPT}{start_time}) {
+                    $node_status_time{$node}{$::STATE_POSTBOOTSCRIPT}{start_time} = $log_ref->{time};
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: could we integrate line 1185 and line 1186?  The same with line 1167 and 1168.

File: xCAT-probe/subcmds/osdeploy [hunk 11]
```diff
@@ -538,13 +551,146 @@ sub conclusion_report {
...
+            push @time_rst, $status_time;
+        }
+    }
+  
+    my $time_str = join("   ", @time_rst);
+    $space = " " x ($length_node - 3);
+    print "ALL $space $time_str\n";
+    probe_utils->send_msg("stdout", "", "------------------------------------------------");
+    probe_utils->send_msg("stdout", "", "$warninfo") if ($warninfo);
+    probe_utils->send_msg("stdout", "", "* : Did not get correct time, please run '$command_input -r xxhxxm' to get correct time") if ($isnull and $is_success);
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: Could we move line 115 to here,  or try to use @ARGV.  It is not a good option to have line 115.

**Comment by hu-weihua** [comment]
Could we change line 450 to 

    ======================  Summary =====================

**Comment by hu-weihua** [comment]
We need detailed comments for function ``performance_calculation``.  I was lost totally.

**Comment by whowutwut** [comment]
@gurevichmark Could you also take a look at this pull request if time permits.

**Comment by gurevichmark** [comment]
1. I think it would be better to use a more meaningful flags, so instead of `-p 1` and `-p 2`, it could be `-p compact` and `-p details`.  
2. Can additional steps of provisioning be shown ? Right now the output shows the elapsed time for DHCP, RPM, POST and POSTBOOT steps. But the numbers for those steps do not add up to COMPLETED time. Users might wonder where the rest of the time was spent.
3. What does ALL line supposed to show ? If `xcatprobe osdeploy` was ran for 2 nodes, provisioning in parallel and each node took 08:55 to boot, will the COMPLETED column on the ALL line show 17:45 min (08:55+08:55)?

**Comment by xuweibj** [comment]
Hi @gurevichmark , 
Thank you for your comments! My answers are as below:

1. I will change the code.
2.Show DHCP, RPM, POST and POSTBOOT is because these steps have start and end time, other steps just have one time point. I think maybe we only can use the time point of next step as the end of current step, but it's not accurate. I don't think it's a good idea.
3. If ran for 2 nodes the COMPLETED column will show the time that from the start of the first node to the end of the last. Not the sum of 2 nodes' time.

**Comment by xuweibj** [comment]
Hi @gurevichmark, I have modified much, please help to review again.
Thanks.

**Comment by gurevichmark** [comment]
Looks good to me.

**Comment by whowutwut** [comment]
@xuweibj  Are we able to use this against a range of nodes, what does the output look like for more than 1 node?

**Comment by xuweibj** [comment]
@whowutwut I updated the comment. The output will like that if 2 nodes.

File: xCAT-probe/subcmds/osdeploy [hunk 12]
```diff
@@ -65,6 +65,11 @@ Options:
     -r : Trigger 'Replay history' mode. Follow the duration of rolling back. Units are 'h' (hour) or 'm' (minute)
          Supported format examples: 3h30m (3 hours and 30 minutes ago), 2h (2 hours ago), 40m (40 minutes ago) and 3 (3 hours ago).
          If unit is not specified, hour will be used by default.
+    -p Show elapsed time of each stage during provision for each node
+         Support 2 output format:
+         'compact': Elapsed time of provision for each node.
+         'phase'  : Elapsed time for DHCP, RPM, POSTSCRIPTS and POSTBOOTSCRIPTS stages, and time for whole provision.
+         'origin' : Show origin time of each stage.
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/osdeploy]: Show origin start time of each stage.

**Comment by gurevichmark** [comment]
Will there still be `ALL` line? Or did you decide to eliminate it ?

**Comment by whowutwut** [comment]
In addition, should  change the column for `RPM` to something else... perhaps `INSTALL`?   rpm is only relevant on OS that uses redhat package manager.

**Comment by xuweibj** [comment]
@gurevichmark we decided to eliminate ``ALL`` line.

@whowutwut , OK. I have changed ``RPM`` to ``INSTALL``.

**Comment by hu-weihua** [comment]
I have reviewed the code change and agree to merge. thanks @xuweibj

**Pull request merged**